devito/petsc/solve.py

```diff
@@ -0,0 +1,116 @@
+                   dimensions=target.space_dimensions,
+                   shape=target.grid.shape,
+                   liveness='eager',
+                   halo=target.halo[1:] if is_time_dep else target.halo)
+        for prefix in ['y_matvec', 'x_matvec', 'b_tmp']]
+
+    # TODO: Extend to rearrange equation for implicit time stepping.
+    matvecaction = MatVecEq(y_matvec, LinearSolveExpr(eq.lhs.subs(target, x_matvec),
+                            target=target, solver_parameters=solver_parameters),
+                            subdomain=eq.subdomain)
+
+    # Part of pde that remains constant at each timestep
+    rhs = RHSEq(b_tmp, LinearSolveExpr(eq.rhs, target=target,
+                solver_parameters=solver_parameters), subdomain=eq.subdomain)
+
+    if not bcs:
+        return [matvecaction, rhs]
+
+    bcs_for_matvec = []
+    for bc in bcs:
+        # TODO: Insert code to distiguish between essential and natural
+        # boundary conditions since these are treated differently within
+        # the solver
+        # NOTE: May eventually remove the essential bcs from the solve
+        # (and move to rhs) but for now, they are included since this
+        # is not trivial to implement when using DMDA
+        # NOTE: Below is temporary -> Just using this as a palceholder for
+        # the actual BC implementation for the matvec callback
+        new_rhs = bc.rhs.subs(target, x_matvec)
+        bc_rhs = LinearSolveExpr(
+            new_rhs, target=target, solver_parameters=solver_parameters
+        )
+        bcs_for_matvec.append(MatVecEq(y_matvec, bc_rhs, subdomain=bc.subdomain))
+
+    return [matvecaction] + bcs_for_matvec + [rhs]
+
+
+def separate_eqn(eqn, target):
+
+    zeroed_eqn = Eq(eqn.lhs - eqn.rhs, 0)
+    tmp = zeroed_eqn.lhs
+
+    tmp2 = eval_time_derivatives(tmp)
+    # Extract the part of the PDE that remains constant at each
+    # time step
+    # In other words, we are solving F(x) = b, where b represents
+    # the part of the pde that doesn't depend on x
+    b = remove_target(tmp2, target)
+    F_x = simplify(tmp2 - b)
+
+    return -b, F_x
+
+
+@singledispatch
+def remove_target(expr, target):
+    return 0 if expr == target else expr
+
+
+@remove_target.register(Add)
+@remove_target.register(EvalDerivative)
+def _(expr, target):
+    if not expr.has(target):
+        return expr
+
+    args = []
+    for a in expr.args:
+        a1 = remove_target(a, target)
+        args.append(a1)
+
+    return expr.func(*args, evaluate=False)
+
+
+@remove_target.register(Mul)
+def _(expr, target):
+    if not expr.has(target):
+        return expr
+
+    args = []
+    for a in expr.args:
+        if not a.has(target):
+            args.append(a)
+        else:
+            a1 = remove_target(a, target)
+            args.append(a1)
+
+    return expr.func(*args, evaluate=False)
+
+
+@remove_target.register(Derivative)
+def _(expr, target):
+    if not expr.has(target):
+        return expr
+    return 0
```

Comment:
20(ish) lines above you use the inline conditional `X if something else Y`, here instead you use a different coding style